### PR TITLE
Ensure queries are closed in API calls

### PR DESCRIPTION
**Pull request by beorn7**
As discussed in #7129.

### CODE REVIEW

**Comment by beorn7** [comment]
One of the many fixes that might help #7120 .

**Comment by brian-brazil** [comment]
I was thinking we might be able to do a defer that only Closed if the named return is missing.

**Comment by beorn7** [comment]
That might work as well. I'll give it a try.

**Comment by beorn7** [comment]
Done. Definitely much nicer.

**Comment by roidelapluie** [comment]
Can you also fix the endpoints which use defer in place of finalizer?

**Comment by beorn7** [comment]
But they don't need fixing, do they?

**Comment by roidelapluie** [comment]
public irc logs
```
20:43:25 <@bbrazil> hmm, series() is using defer - that may cause memory corruption
20:43:41 <@bbrazil> we need to keep the strings referenced until the api call is complete
20:44:00 <@bbrazil> well, a memory exposure to be more precise
20:44:48 < roidelapluie> series, labels, label_values..
20:49:22 <@bbrazil> not sure how I missed those, because I know I checked when making those changes
20:49:47 <@bbrazil> maybe they are safe, but the tsdb internals will only get more mmap-y so may aswell make them future proof now
```

**Comment by roidelapluie** [comment]
> But they don't need fixing, do they?

Yeah we can analyze that deeper later.

**Comment by beorn7** [comment]
I see. I read that discussion, but didn't connect it to _this_ `series` call. I'll leave that to the TSDB experts. (o:

**Comment by brian-brazil** [comment]
:+1: 

Yeah, `series` needs the same pattern.

**Comment by beorn7** [comment]
Then let me add that quickly.

File: web/api/v1/api.go [hunk 1]
```diff
@@ -549,15 +549,25 @@ func (api *API) series(r *http.Request) apiFuncResult {
...
+	}
 
 	var sets []storage.SeriesSet
 	var warnings storage.Warnings
 	for _, mset := range matcherSets {
 		s, wrn, err := q.Select(false, nil, mset...)
 		warnings = append(warnings, wrn...)
 		if err != nil {
-			return apiFuncResult{nil, &apiError{errorExec, err}, warnings, nil}
+			return apiFuncResult{nil, &apiError{errorExec, err}, warnings, closer}
```
Comment (roidelapluie) [inline web/api/v1/api.go:570]:
```suggestion
			return apiFuncResult{nil, &apiError{errorExec, err}, warnings, q.Close}
```

File: web/api/v1/api.go [hunk 2]
```diff
@@ -568,10 +578,10 @@ func (api *API) series(r *http.Request) apiFuncResult {
 		metrics = append(metrics, set.At().Labels())
 	}
 	if set.Err() != nil {
-		return apiFuncResult{nil, &apiError{errorExec, set.Err()}, warnings, nil}
+		return apiFuncResult{nil, &apiError{errorExec, set.Err()}, warnings, closer}
```
Comment (roidelapluie) [inline web/api/v1/api.go:581]:
```suggestion
		return apiFuncResult{nil, &apiError{errorExec, set.Err()}, warnings, q.Close}
```

File: web/api/v1/api.go [hunk 3]
```diff
@@ -568,10 +578,10 @@ func (api *API) series(r *http.Request) apiFuncResult {
...
 	}
 
-	return apiFuncResult{metrics, nil, warnings, nil}
+	return apiFuncResult{metrics, nil, warnings, closer}
```
Comment (roidelapluie) [inline web/api/v1/api.go:584]:
```suggestion
	return apiFuncResult{metrics, nil, warnings, q.Close}
```

Comment (beorn7) [reply web/api/v1/api.go:584]: Computer says no.

Or more precisely: “cannot use q.Close (type func() error) as type func() in field value”

Comment (roidelapluie) [reply web/api/v1/api.go:584]: ooh ok.

**Comment by roidelapluie** [comment]
LGTM

**Comment by brian-brazil** [comment]
:+1: 

Thanks

**Pull request merged**